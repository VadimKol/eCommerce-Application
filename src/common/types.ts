import type { BaseAddress, CustomFieldsDraft, StoreResourceIdentifier } from '@commercetools/platform-sdk';

export interface AuthContextInterface {
  isAuthenticated: boolean;
  handleLogin: () => void;
  handleLogout: () => void;
}

export type CategoriesContextType = {
  categories: CategoriesData | null;
  loading: boolean;
  error: Error | null;
};

export interface Credentials {
  clientId: string;
  clientSecret: string;
}

export interface User {
  username: string;
  password: string;
}

export interface GeekShopCustomerDraft {
  /**
   *	Email address of the Customer that is [unique](/../api/customers-overview#customer-uniqueness) for an entire Project or Store the Customer is assigned to.
   *	It is the mandatory unique identifier of a Customer.
   *
   *
   */
  readonly email: string;
  /**
   *	Password of the Customer.
   *
   *
   */
  readonly password: string;
  /**
   *	Given name (first name) of the Customer.
   *
   *
   */
  readonly firstName?: string;
  /**
   *	Family name (last name) of the Customer.
   *
   *
   */
  readonly lastName?: string;
  /**
   *	Middle name of the Customer.
   *
   *
   */
  readonly middleName?: string;
  /**
   *	Title of the Customer, for example, 'Dr.'.
   *
   *
   */
  readonly title?: string;
  /**
   *	Salutation of the Customer, for example, 'Mr.' or 'Mrs.'.
   *
   *
   */
  readonly salutation?: string;
  /**
   *	Date of birth of the Customer.
   *
   *
   */
  readonly dateOfBirth?: string;
  /**
   *	Company name of the Customer.
   *
   *
   */
  readonly companyName?: string;
  /**
   *	Individual VAT ID of the Customer.
   *
   *
   */
  readonly vatId?: string;
  /**
   *	Addresses of the Customer.
   *
   *
   */
  readonly addresses?: BaseAddress[];
  /**
   *	Indexes of entries in `addresses` to set as shipping addresses.
   *	The `shippingAddressIds` of the [Customer](ctp:api:type:Customer) will be replaced by these addresses.
   *
   *
   */
  readonly shippingAddresses?: number[];
  /**
   *	Index of the address in the `addresses` array to use as the default shipping address.
   *	The `defaultShippingAddressId` of the Customer will be set to the `id` of that address.
   *
   *
   */
  readonly defaultShippingAddress?: number;
  /**
   *	Indexes of entries in `addresses` to set as billing addresses.
   *	The `billingAddressIds` of the [Customer](ctp:api:type:Customer) will be replaced by these addresses.
   *
   *
   */
  readonly billingAddresses?: number[];
  /**
   *	Index of the address in the `addresses` array to use as the default billing address.
   *	The `defaultBillingAddressId` of the Customer will be set to the `id` of that address.
   *
   *
   */
  readonly defaultBillingAddress?: number;
  /**
   *	Custom Fields for the Customer.
   *
   *
   */
  readonly custom?: CustomFieldsDraft;
  /**
   *	Preferred language of the Customer. Must be one of the languages supported by the [Project](ctp:api:type:Project).
   *
   *
   */
  readonly locale?: string;
  /**
   *	Sets the [Stores](ctp:api:type:Store) for the Customer.
   *
   *
   */
  readonly stores?: StoreResourceIdentifier[];
}

export type ProductDetails = {
  name?: string;
  description?: string;
  price?: string | null;
  discountedPrice?: string | null;
  currency?: string | null;
  images?: string[];
  attributes?: { name: string; value: string | number | string[] }[];
  availability?: {
    isOnStock?: boolean | null;
    availableQuantity?: number | null;
  };
};

export type Product = {
  id: string;
  name: string;
  description: string;
  slug: string;
  key: string;
  sku: string;
  quantity: number;
  price: string;
  images: string[];
  categoryId: string;
  slugCategory: string;
  keyCategory: string;
  subcategoryId: string;
  slugSubCategory: string;
  keySubCategory: string;
};

export type Subcategory = {
  id: string;
  name: string;
  slug: string;
  key: string;
  products: Product[];
};

export type Category = {
  id: string;
  name: string;
  slug: string;
  key: string;
  subcategories: Subcategory[];
};

export type CategoriesData = Category[];

export type UrlParams = {
  categoryName?: string;
  subcategoryName?: string;
  productName?: string;
};

export type LoaderData = {
  category?: Category;
  subcategory?: Subcategory;
  product?: Product;
};
