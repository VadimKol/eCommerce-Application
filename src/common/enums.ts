export enum NavigationPaths {
  HOME = '/',
  CATALOG = '/catalog',
  ABOUT = '/about',
}

export enum ActionPaths {
  LOGIN = '/login',
  REGISTER = '/register',
  PROFILE = '/profile',
}
