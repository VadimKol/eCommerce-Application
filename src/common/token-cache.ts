import type { TokenCache, TokenStore } from '@commercetools/sdk-client-v2';

export class CustomTokenCache implements TokenCache {
  private tokenStore: TokenStore = {
    token: localStorage.getItem('geek-shop-token') ? (localStorage.getItem('geek-shop-token') as string) : '',
    expirationTime: 0,
    refreshToken: '',
  };

  public set(tokenStore: TokenStore): void {
    this.tokenStore = tokenStore;
  }

  public get(): TokenStore {
    return this.tokenStore;
  }
}
