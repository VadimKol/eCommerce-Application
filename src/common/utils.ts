import type { CatalogAction, CatalogState } from './types';

export const calculateAge = (birthDate: string): boolean => {
  const today = new Date();
  const birth = new Date(birthDate);
  let age = today.getFullYear() - birth.getFullYear();
  const monthDifference = today.getMonth() - birth.getMonth();
  if (monthDifference < 0 || (monthDifference === 0 && today.getDate() < birth.getDate())) {
    age -= 1;
  }
  return age >= 13;
};

export const sortingTypes = ['Alphabetically, A-Z', 'Alphabetically, Z-A', 'Price, low to high', 'Price, high to low'];

export const fandoms = [
  'Genshin',
  'Disney',
  'Cyberpunk',
  'Batman',
  'Doctor Who',
  'Star Wars',
  'Beetlejuice',
  'Marvel',
  'One Piece',
  'Harry Potter',
  'Stranger Things',
  'Rick and Morty',
  'Supernatural',
];

export const countriesFilter = ['China', 'France', 'Germany', 'India', 'Italy', 'Korea', 'US'];
export const materials = ['Aluminum', 'Cardboard', 'Ceramic', 'Coir', 'Paper', 'Plastic', 'Polyester'];
export const countries = [
  {
    title: 'United States',
    code: 'US',
    errorMsg: 'It must be 5 digits or 5+4 digits',
    postcode: /^\d{5}(-\d{4})?$/,
  },
  {
    title: 'Belarus',
    code: 'BY',
    errorMsg: 'It must be exactly 6 digits.',
    postcode: /^\d{6}$/,
  },
  {
    title: 'Russia',
    code: 'RU',
    errorMsg: 'It must be exactly 6 digits.',
    postcode: /^\d{6}$/,
  },
];

export const DEFAULT_LOCALE = 'en-US';
export const QUERY_LIMIT = 12;
export const PRICE_FILTER_MIN = 1;
export const PRICE_FILTER_MAX = 100;
export const REFRESH_TOKEN_EXPIRATION_DAYS = 150;
export const Discounts = [
  {
    name: 'GEEK-SHOP',
    desc: '-10% cart',
  },
  {
    name: 'COSPLAY',
    desc: '-10% category',
  },
];

export function getCheckboxFilter(checkboxes: boolean[], attribute: string): string {
  const filter: string[] = [];
  let constantsFilter: string[] = [];

  switch (attribute) {
    case 'Fandom': {
      constantsFilter = fandoms;
      break;
    }
    case 'Country': {
      constantsFilter = countriesFilter;
      break;
    }
    case 'Material': {
      constantsFilter = materials;
      break;
    }
    default:
      break;
  }

  checkboxes.forEach((checkbox, index) => {
    constantsFilter.forEach((element, i) => {
      if (index === i && checkbox) {
        filter.push(`"${element}"`);
      }
    });
  });

  return filter.length > 0 ? `variants.attributes.${attribute}:${filter.join(',')}` : '';
}

export function getSort(sortType: string): string | undefined {
  switch (sortType) {
    case 'Alphabetically, A-Z':
      return `name.${DEFAULT_LOCALE} asc`;
    case 'Alphabetically, Z-A':
      return `name.${DEFAULT_LOCALE} desc`;
    case 'Price, low to high':
      return 'price asc';
    case 'Price, high to low':
      return 'price desc';
    default:
      return undefined;
  }
}

export class StatusError extends Error {
  public statusCode: number;

  constructor(message: string, status: number) {
    super(message);
    this.statusCode = status;
  }
}

export const initialState: CatalogState = {
  products: [],
  page: 0,
  sortType: 'Sort',
  search: '',
  priceFilter: [PRICE_FILTER_MIN, PRICE_FILTER_MAX],
  franchises: Array(fandoms.length).fill(false),
  countriesF: Array(countriesFilter.length).fill(false),
  materialsF: Array(materials.length).fill(false),
  loadingProducts: true,
  categories: {},
};

export function reducerCatalog(
  state: CatalogState,
  {
    type,
    products = [],
    page = 0,
    sortType = 'Sort',
    search = '',
    priceFilter = [PRICE_FILTER_MIN, PRICE_FILTER_MAX],
    franchises = Array(fandoms.length).fill(false),
    countriesF = Array(countriesFilter.length).fill(false),
    materialsF = Array(materials.length).fill(false),
    loadingProducts = true,
    categories = {},
  }: CatalogAction,
): CatalogState {
  switch (type) {
    case 'SET_CATEGORIES':
      return {
        ...state,
        categories,
        page,
        sortType,
        priceFilter,
        franchises,
        countriesF,
        materialsF,
        search,
        loadingProducts,
      };
    case 'SET_PRODUCTS':
      return { ...state, products };
    case 'SET_LOADING_PRODUCTS':
      return { ...state, loadingProducts };
    case 'SET_SEARCH':
      return { ...state, search, page, loadingProducts };
    case 'SET_SORT_TYPE':
      return { ...state, sortType, loadingProducts };
    case 'SET_PRICE':
      return { ...state, priceFilter, page, loadingProducts };
    case 'SET_FRANCHISES':
      return { ...state, franchises, page, loadingProducts };
    case 'SET_COUNTRIES':
      return { ...state, countriesF, page, loadingProducts };
    case 'SET_MATERIALS':
      return { ...state, materialsF, page, loadingProducts };
    case 'RESET_FILTERS':
      return { ...state, priceFilter, franchises, countriesF, materialsF, page, loadingProducts };
    case 'SET_PAGE':
      return { ...state, page, loadingProducts };
    default:
      return state;
  }
}

export const capitalizeFirstLetter = (string: string): string => string.charAt(0).toUpperCase() + string.slice(1);
