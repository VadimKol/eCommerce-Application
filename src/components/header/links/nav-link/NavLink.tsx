import classNames from 'classnames';
import { Link, useLocation } from 'react-router-dom';

import styles from './styles.module.scss';

interface NavLinkProps {
  to: string;
  label: string;
  className?: string;
  setIsCurrent?: boolean;
}

export function NavLink({ to, label, className, setIsCurrent = false }: NavLinkProps): JSX.Element {
  const location = useLocation();
  const isCurrentLink = location.pathname === to;

  if (isCurrentLink) {
    return <span className={classNames(className, styles.link, styles.linkCurrent)}>{label}</span>;
  }

  return (
    <Link className={classNames(className, styles.link, { [styles.linkCurrent!]: setIsCurrent })} to={to}>
      {label}
    </Link>
  );
}
