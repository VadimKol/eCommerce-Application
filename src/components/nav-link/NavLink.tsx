import classNames from 'classnames';
import { Link, useLocation } from 'react-router-dom';

import styles from './styles.module.scss';

interface NavLinkProps {
  to: string;
  label: string;
  className?: string;
  setIsCurrent?: boolean;
  onClick?: React.MouseEventHandler;
  icon?: string;
  extraInfo?: string;
  insideBurgerMenu?: boolean;
}

export function NavLink({
  to,
  label,
  className,
  setIsCurrent = false,
  onClick,
  icon = '',
  extraInfo = '',
  insideBurgerMenu = false,
}: NavLinkProps): JSX.Element {
  const location = useLocation();
  const isCurrentLink = location.pathname === to;

  if (isCurrentLink && !icon) {
    return <span className={classNames(className, styles.link, styles.linkCurrent)}>{label}</span>;
  }

  if (isCurrentLink && icon) {
    return (
      <span
        className={classNames(
          className,
          styles.link,
          styles.linkCurrent,
          styles.linkIcon,
          styles[`linkIcon--${icon}`],
          { [styles.linkExtra || '']: extraInfo },
          { [styles.insideBurgerMenu || '']: insideBurgerMenu },
        )}
      >
        <span className="visually-hidden">{label}</span>
        {extraInfo && <span className={styles.extraInfo}>{extraInfo}</span>}{' '}
      </span>
    );
  }

  return (
    <Link
      className={classNames(
        className,
        styles.link,
        { [styles.linkCurrent || '']: setIsCurrent },
        { [styles.linkIcon || '']: icon },
        { [styles[`linkIcon--${icon}`] || '']: icon },
        { [styles.linkExtra || '']: extraInfo },
        { [styles.insideBurgerMenu || '']: insideBurgerMenu },
      )}
      to={to}
      onClick={onClick}
    >
      <span className={icon ? 'visually-hidden' : ''}>{label}</span>
      {extraInfo && <span className={styles.extraInfo}>{extraInfo}</span>}{' '}
    </Link>
  );
}
