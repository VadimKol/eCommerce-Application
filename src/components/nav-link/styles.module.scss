
.link {
  display: flex;
  gap: 10px;
  justify-content: center;
  align-self: center;
  padding: 5px;
  font-size: inherit;
  text-decoration: none;
  color: var(--text-color);
  transition: color 0.3s ease-in-out, text-decoration-color 0.3s ease-in-out;

  &:hover,
  &:focus-visible {
    @media (hover: hover) {
      color: var(--tertiary-color);
      text-decoration-color: var(--tertiary-color);
    }
  }
}

.extraInfo {
  z-index: 1;
  grid-area: 1 / 1 / 2 / 2;
  min-width: 27px;
  padding: 2px 4px;
  font-size: 16px;
  text-align: center;
  background-image: linear-gradient(to bottom, #0b7428, #0d812e);
  border-radius: 14px;

  .insideBurgerMenu & {
    background-image: linear-gradient(to bottom, var(--primary-color), #002974);
  }
}

.linkExtra {
  display: grid;

  &::after {
    grid-area: 1 / 1 / 2 / 2;
    align-self: center;
  }

  & .extraInfo {
    align-self: start;
    margin: -10px 0 10px 16px;
  }
}

.linkIcon {
  &::after {
    content: "";
    width: 30px;
    height: 30px;
    background-color: var(--text-color);
    transition: background-color 0.3s ease-in-out;
    mask-size: cover;
    mask-repeat: no-repeat;
    mask-position: center;
  }

  &:hover,
  &:focus-visible {
    &::after {
      background-color: var(--tertiary-color);
    }
  }

  &--cart {
    &::after {
      mask-image: url("/src/assets/images/cart.svg");
    }
  }
}

.linkCurrent {
  text-decoration: underline;
  text-underline-offset: 10px;
  text-decoration-thickness: 2px;
  text-decoration-color: var(--text-color);

  &::after {
    background-color: var(--tertiary-color);
  }

  &.linkExtra {
    text-decoration: none;

    & .extraInfo {
      color: var(--tertiary-color);
    }
  }
}

span.linkCurrent {
  pointer-events: none;
}
