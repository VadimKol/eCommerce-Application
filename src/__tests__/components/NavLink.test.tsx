import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { ActionPaths } from '@/common/enums';
import { NavLink } from '@/components/nav-link/NavLink';

it('renders correctly', () => {
  const { container } = render(
    <BrowserRouter>
      <NavLink to={ActionPaths.CART} label="Items in cart:" icon="cart" />
    </BrowserRouter>,
  );

  expect(container).toMatchSnapshot();
});
