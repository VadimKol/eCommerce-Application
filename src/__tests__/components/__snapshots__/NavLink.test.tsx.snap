// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <a
    class="link linkIcon linkIcon--cart"
    href="/cart"
  >
    <span
      class="visually-hidden"
    >
      Items in cart:
    </span>
     
  </a>
</div>
`;
